src/ast_selectors.cpp

```diff
@@ -363,1113 +312,719 @@ namespace Sass {
-            Placeholder_Selector* pss = SASS_MEMORY_COPY(ps);
-            pss->name(pss->name() + (*h)[0]->name());
-            pss->pstate((*h)[0]->pstate());
-            (*rh)[rh->length()-1] = pss;
-            rh->pstate(h->pstate());
-            for (i = 1; i < L; ++i) rh->append((*h)[i]);
-          } else {
-            last()->head_->concat(h);
-          }
-        } else {
-          last()->head_->concat(h);
-        }
-      } else if (last()->head_) {
-        last()->head_->concat(h);
-      }
-    } else {
-      // std::cerr << "has no or empty head\n";
-    }
+  CssMediaRule::CssMediaRule(ParserState pstate, Block_Obj block) :
+    Has_Block(pstate, block),
+    Vectorized()
+  {
+    statement_type(MEDIA);
+  }
 
-    Complex_Selector* last = mutable_last();
-    if (last) {
-      if (last->combinator() != ANCESTOR_OF && c != ANCESTOR_OF) {
-        Complex_Selector* inter = SASS_MEMORY_NEW(Complex_Selector, pstate());
-        inter->reference(r);
-        inter->combinator(c);
-        inter->tail(t);
-        last->tail(inter);
-      } else {
-        if (last->combinator() == ANCESTOR_OF) {
-          last->combinator(c);
-          last->reference(r);
-        }
-        last->tail(t);
-      }
-    }
+  CssMediaRule::CssMediaRule(const CssMediaRule* ptr) :
+    Has_Block(ptr),
+    Vectorized(*ptr)
+  {
+    statement_type(MEDIA);
+  }
 
+  CssMediaQuery::CssMediaQuery(ParserState pstate) :
+    AST_Node(pstate),
+    modifier_(""),
+    type_(""),
+    features_()
+  {
   }
 
-  Selector_List* Complex_Selector::resolve_parent_refs(SelectorStack& pstack, Backtraces& traces, bool implicit_parent)
+  /////////////////////////////////////////////////////////////////////////
+  /////////////////////////////////////////////////////////////////////////
+
+  bool CssMediaQuery::operator==(const CssMediaQuery& rhs) const
   {
-    Complex_Selector_Obj tail = this->tail();
-    Compound_Selector_Obj head = this->head();
-    Selector_List* parents = pstack.back();
+    return type_ == rhs.type_
+      && modifier_ == rhs.modifier_
+      && features_ == rhs.features_;
+  }
 
-    if (!this->has_real_parent_ref() && !implicit_parent) {
-      Selector_List* retval = SASS_MEMORY_NEW(Selector_List, pstate(), 1);
-      retval->append(this);
-      return retval;
-    }
+  // Implemented after dart-sass (maybe move to other class?)
+  CssMediaQuery_Obj CssMediaQuery::merge(CssMediaQuery_Obj& other)
+  {
 
-    // first resolve_parent_refs the tail (which may return an expanded list)
-    Selector_List_Obj tails = tail ? tail->resolve_parent_refs(pstack, traces, implicit_parent) : 0;
+    std::string ourType(this->type());
+    std::string theirType(other->type());
+    std::string ourModifier(this->modifier());
+    std::string theirModifier(other->modifier());
 
-    if (head && head->length() > 0) {
+    std::transform(ourType.begin(), ourType.end(), ourType.begin(), ::tolower);
```

Comment:
We shouldn't use `tolower`, because it's C-locale dependent. I'll send a PR shortly adding our own method.

Reply:
Can we fix this after this PR has landed?

Reply:
Absolutely